lib/auth/version.go

```diff
@@ -20,55 +20,136 @@ package auth
+	} else if err != nil {
+		return trace.Wrap(err)
 	}
 
-	lastKnownVersion, err := storage.GetTeleportVersion(ctx)
-	if trace.IsNotFound(err) {
-		if err := storage.WriteTeleportVersion(ctx, currentVersion); err != nil {
+	var createNewResource bool
+	for range versionUpgradeCheckMaxWriteRetry {
+		backendInfo, err := backendStorage.GetBackendInfo(ctx)
+		if trace.IsNotFound(err) {
+			createNewResource = true
+			backendInfo, err = backendinfo.NewBackendInfo(&backendinfov1.BackendInfoSpec{TeleportVersion: teleportVersion.String()})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		} else if err != nil {
 			return trace.Wrap(err)
 		}
+
+		lastKnownVersion, err := semver.NewVersion(backendInfo.GetSpec().GetTeleportVersion())
+		if err != nil {
+			return trace.Wrap(err, "failed to parse teleport version: %+q", backendInfo.GetSpec().GetTeleportVersion())
+		}
+		// The last known version is already updated to the current one.
+		// Skip any further checks and resource updates.
+		if !createNewResource && lastKnownVersion.Equal(currentVersion) {
+			return nil
+		}
+		if currentVersion.Major-lastKnownVersion.Major > 1 {
+			return trace.BadParameter("Unsupported upgrade path detected: from %v to %v. "+
+				"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major+1)
+		}
+		if lastKnownVersion.Major > currentVersion.Major {
+			return trace.BadParameter("Unsupported downgrade path detected: from %v to %v. "+
+				"Teleport doesn't support major version downgrade.\n Please downgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
+				lastKnownVersion, currentVersion.String(), lastKnownVersion.Major-1)
+		}
+
+		backendInfo.GetSpec().TeleportVersion = currentVersion.String()
+
+		if createNewResource {
+			_, err = backendStorage.CreateBackendInfo(ctx, backendInfo)
+			if trace.IsAlreadyExists(err) {
+				err = trace.Wrap(err)
+				slog.WarnContext(ctx, "Failed to create BackendInfo resource", "error", err)
+				continue
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+		} else {
+			_, err = backendStorage.UpdateBackendInfo(ctx, backendInfo)
+			if errors.Is(err, backend.ErrIncorrectRevision) || trace.IsNotFound(err) {
+				err = trace.Wrap(err)
+				slog.WarnContext(ctx, "Failed to update BackendInfo resource", "error", err)
+				continue
+			} else if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		// TODO(vapopov): DELETE IN v19.0.0 – the last known version should already be migrated to backend storage.
+		if cleanProcVersion {
+			if err := procStorage.DeleteTeleportVersion(ctx); err != nil {
+				return trace.Wrap(err)
+			}
```

Comment:
extreme nit just in case we have an early restart of the auth and two concurrent processes on the same process storage:
```suggestion
			if err := procStorage.DeleteTeleportVersion(ctx); err != nil && !trace.IsNotFound(err) {
				return trace.Wrap(err)
			}
```